ambry-router/src/main/java/com.github.ambry.router/PutManager.java

```diff
@@ -83,18 +91,35 @@ public void registerRequestToSend(PutOperation putOperation, RequestInfo request
    * @param routerConfig  The {@link RouterConfig} containing the configs for the PutManager.
    * @param routerMetrics The {@link NonBlockingRouterMetrics} to be used for reporting metrics.
    * @param operationCompleteCallback The {@link OperationCompleteCallback} to use to complete operations.
+   * @param readyForPollCallback The callback to be used to notify the router of any state changes within the
+   *                             operations.
    * @param index the index of the {@link NonBlockingRouter.OperationController} in the {@link NonBlockingRouter}
    * @param time The {@link Time} instance to use.
    */
   PutManager(ClusterMap clusterMap, ResponseHandler responseHandler, NotificationSystem notificationSystem,
       RouterConfig routerConfig, NonBlockingRouterMetrics routerMetrics,
-      OperationCompleteCallback operationCompleteCallback, int index, Time time) {
+      OperationCompleteCallback operationCompleteCallback, ReadyForPollCallback readyForPollCallback, int index,
+      Time time) {
     this.clusterMap = clusterMap;
     this.responseHandler = responseHandler;
     this.notificationSystem = notificationSystem;
     this.routerConfig = routerConfig;
     this.routerMetrics = routerMetrics;
     this.operationCompleteCallback = operationCompleteCallback;
+    this.readyForPollCallback = readyForPollCallback;
+    this.chunkArrivalListener = new ByteBufferAsyncWritableChannel.ChannelEventListener() {
+      @Override
+      public void onEvent(ByteBufferAsyncWritableChannel.EventType e) {
+        synchronized (chunkFillerSynchronizer) {
+          // At this point, the chunk for which this notification came in (if any) could already have been consumed by
+          // the chunk filler, and this might unnecessarily wake it up from its sleep, which should be okay.
+          areChunksAvailableForFilling = true;
+          if (isChunkFillerThreadAsleep) {
```

Comment:
I am wondering how necessary this check is. Can we just notify the `chunkfiller` when some data is ready from the async channel?